File: tester/bench_tester.py
import psutil
import time
import os

def monitor_usage(duration):
    """Monitorea el uso de CPU y memoria cada segundo durante la ejecución."""
    cpu_usage = []
    memory_usage = []

    for _ in range(duration):
        cpu_usage.append(psutil.cpu_percent(interval=1))
        memory_usage.append(psutil.virtual_memory().percent)

    return cpu_usage, memory_usage

def execute_and_monitor(command):
    """Ejecuta un comando y monitorea el uso del sistema."""
    start_time = time.time()

    # Inicia el proceso
    for i in range(1,30):
        process = os.system(command)

    # Monitoreo en paralelo
    duration = int(time.time() - start_time)
    cpu_usage, memory_usage = monitor_usage(duration)

    end_time = time.time()
    execution_time = end_time - start_time

    return execution_time, cpu_usage, memory_usage

    

def main():
    # Comando para ejecutar el algoritmo
    command = "python3 bike-t.py"

    # Ejecutar y monitorear
    execution_time, cpu_usage, memory_usage = execute_and_monitor(command)

    # Resultados
    print(f"Tiempo de ejecución: {execution_time:.2f} segundos")
    print(f"Uso promedio de CPU: {sum(cpu_usage) / len(cpu_usage):.2f}%")
    print(f"Uso promedio de memoria: {sum(memory_usage) / len(memory_usage):.2f}%")


if __name__=="__main__":
    main()

File: tester/bench_tester.cpp
#include <iostream>
#include <fstream>
#include <chrono>
#include <thread>
#include <cstdlib>
#include <string>
#include <sstream>

// Función para leer el uso de CPU del sistema
float getCPUUsage() {
    std::ifstream statFile("/proc/stat");
    std::string line, cpu;
    long user, nice, system, idle, iowait, irq, softirq, steal;
    if (statFile.is_open()) {
        std::getline(statFile, line);
        std::istringstream ss(line);
        ss >> cpu >> user >> nice >> system >> idle >> iowait >> irq >> softirq >> steal;
        statFile.close();
        long idleTime = idle + iowait;
        long totalTime = user + nice + system + idle + iowait + irq + softirq + steal;
        return (1.0 - (float(idleTime) / totalTime)) * 100.0;
    }
    return 0.0; // Retorna 0 si no se pudo leer
}

// Función para leer el uso de memoria del sistema
float getMemoryUsage() {
    std::ifstream meminfoFile("/proc/meminfo");
    std::string line, key;
    long totalMemory = 0, freeMemory = 0;
    if (meminfoFile.is_open()) {
        while (std::getline(meminfoFile, line)) {
            std::istringstream ss(line);
            ss >> key;
            if (key == "MemTotal:") {
                ss >> totalMemory;
            } else if (key == "MemAvailable:") {
                ss >> freeMemory;
                break;
            }
        }
        meminfoFile.close();
        return (1.0 - (float(freeMemory) / totalMemory)) * 100.0;
    }
    return 0.0; // Retorna 0 si no se pudo leer
}

int main() {
    // Comando a ejecutar
    std::string command = "python3 pqc_algorithm.py";

    // Variables para monitoreo
    float totalCPUUsage = 0.0;
    float totalMemoryUsage = 0.0;
    int samples = 0;

    // Inicio del temporizador
    auto start = std::chrono::high_resolution_clock::now();

    // Ejecutar el comando en un hilo separado
    std::thread commandThread([&]() {
        system(command.c_str());
    });

    // Monitoreo durante la ejecución
    while (commandThread.joinable()) {
        float cpuUsage = getCPUUsage();
        float memoryUsage = getMemoryUsage();

        totalCPUUsage += cpuUsage;
        totalMemoryUsage += memoryUsage;
        samples++;

        std::this_thread::sleep_for(std::chrono::seconds(1)); // Tomar muestra cada segundo
    }

    // Final del temporizador
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> executionTime = end - start;

    // Esperar a que termine el hilo
    commandThread.join();

    // Cálculo de promedios
    float avgCPUUsage = totalCPUUsage / samples;
    float avgMemoryUsage = totalMemoryUsage / samples;

    // Imprimir resultados
    std::cout << "Tiempo de ejecución: " << executionTime.count() << " segundos\n";
    std::cout << "Uso promedio de CPU: " << avgCPUUsage << "%\n";
    std::cout << "Uso promedio de memoria: " << avgMemoryUsage << "%\n";

    // Guardar resultados en un archivo
    std::ofstream outputFile("usage_metrics.txt");
    outputFile << "Tiempo de ejecución: " << executionTime.count() << " segundos\n";
    outputFile << "Uso promedio de CPU: " << avgCPUUsage << "%\n";
    outputFile << "Uso promedio de memoria: " << avgMemoryUsage << "%\n";
    outputFile.close();

    return 0;
}